### PR TITLE
Harnesses verifying slice types for add, sub and offset

### PR DESCRIPTION
**Pull request by szlee118**
Towards https://github.com/model-checking/verify-rust-std/issues/76

### Contracts Added:

- Added contracts for <*mut T>::add, <*mut T>::sub, and <*mut T>::offset operations for raw pointers to slices.
### Proofs Implemented:

- Added proofs for the above contracts, verifying safety for the following slice pointee types using arrays:
 - All integer types (i8, i16, i32, etc.).
 - Tuples (composite types) (e.g., (i8, i8), (i32, f64, bool)).
 - Unit type (()).
 
### Macro Definitions:
- Introduced macros to simplify and automate harness generation:
 - One macro for add and sub operations on slices.
 - Another macro for offset operations on slices.
 
### Array-Based Implementation for Slice Verification:
- Arrays are used as a proxy for slices (as slices do not have a known length at compile time) to enable proper bounds checking and assumptions during verification.

### This should be merged after [another PR](https://github.com/model-checking/verify-rust-std/pull/113/files) for dependency issues.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

**Comment by szlee118** [comment]
@feliperodri FYI

**Comment by tautschnig** [comment]
@szlee118 #113 has been merged, but now conflicts need to be resolved on this one.

**Comment by szlee118** [comment]
@tautschnig @feliperodri conflict resolved.

File: library/core/src/ptr/mut_ptr.rs [hunk 1]
```diff
@@ -2415,13 +2528,15 @@ mod verify {
     generate_mut_arithmetic_harness!(i64, check_mut_add_i64, add);
     generate_mut_arithmetic_harness!(i128, check_mut_add_i128, add);
     generate_mut_arithmetic_harness!(isize, check_mut_add_isize, add);
+    generate_mut_arithmetic_harness!(u8, check_mut_add_u8, add);
```
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2531]: If we add this line, the comment below should be removed and the issue solved.

File: library/core/src/ptr/mut_ptr.rs [hunk 2]
```diff
@@ -2415,13 +2528,15 @@ mod verify {
...
     // Due to a bug of kani this test case is malfunctioning for now.
     // Tracking issue: https://github.com/model-checking/kani/issues/3743
     // generate_mut_arithmetic_harness!(u8, check_mut_add_u8, add);
     generate_mut_arithmetic_harness!(u16, check_mut_add_u16, add);
     generate_mut_arithmetic_harness!(u32, check_mut_add_u32, add);
     generate_mut_arithmetic_harness!(u64, check_mut_add_u64, add);
     generate_mut_arithmetic_harness!(u128, check_mut_add_u128, add);
+    generate_mut_arithmetic_harness!(usize, check_mut_add_usize, add);
```
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2539]: duplicated line

File: library/core/src/ptr/mut_ptr.rs [hunk 3]
```diff
@@ -2454,6 +2569,7 @@ mod verify {
     generate_mut_arithmetic_harness!((i8, i8), check_mut_sub_tuple_1, sub);
     generate_mut_arithmetic_harness!((f64, bool), check_mut_sub_tuple_2, sub);
     generate_mut_arithmetic_harness!((i32, f64, bool), check_mut_sub_tuple_3, sub);
+    generate_mut_arithmetic_harness!((i8, u16, i32, u64, isize), check_mut_sub_tuple_4, sub);
```
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2572]: duplicated line

**Review by feliperodri — COMMENTED** [review]
It seems there are still some conflicts. Could you check if the conflicts were successfully solved?

File: library/core/src/ptr/mut_ptr.rs (hunk 1 above)
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Line removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs (hunk 3 above)
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs:2570]: Line removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs (hunk 2 above)
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Line removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs [hunk 4]
```diff
@@ -2370,6 +2450,17 @@ mod verify {
         ($type:ty, $proof_name:ident, add) => {
             #[kani::proof_for_contract(<*mut $type>::add)]
             pub fn $proof_name() {
+                let mut test_val: $type = kani::any::<$type>();
+                let offset: usize = kani::any();
+                let count: usize = kani::any();                
+                let test_ptr: *mut $type = &mut test_val;
+
+                // For integer, composite, and unit types, 1 is the largest offset that
+                // keeps `ptr_with_offset` within bounds.
```
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2459]: It's missing the key part of the explanation: why? why 1 keeps `ptr_with_offset` for these types?

File: library/core/src/ptr/mut_ptr.rs [hunk 5]
```diff
@@ -2361,6 +2367,80 @@ impl<T: ?Sized> PartialOrd for *mut T {
 #[unstable(feature = "kani", issue = "none")]
 mod verify {
     use crate::kani;
+    use core::mem;
+    // Constant for array size used in all tests, for performance reason
+    const ARRAY_SIZE: usize = 5;
```
Comment (feliperodri) [inline library/core/src/ptr/mut_ptr.rs:2372]: Isn't 5 too low? How is the verification performance numbers?

Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs:2372]: Explained in another [PR's comment](https://github.com/model-checking/verify-rust-std/pull/166#discussion_r1861482793) that 5 is enough for this case.
This `performance` statement is outdated, and has been removed by newest commit.

File: library/core/src/ptr/mut_ptr.rs (hunk 4 above)
Comment (xsxszab) [reply library/core/src/ptr/mut_ptr.rs:2459]: This is a legacy implementation for the proof of contract. In our latest implementation (detailed in PR #166), it has been updated to use the pointer generator.

**Comment by szlee118** [comment]
> It seems there are still some conflicts. Could you check if the conflicts were successfully solved?

It has been solved successfully.

**Review by feliperodri — APPROVED** [review]

File: library/core/src/ptr/mut_ptr.rs [hunk 6]
```diff
@@ -404,8 +404,10 @@ impl<T: ?Sized> *mut T {
     #[rustc_const_stable(feature = "const_ptr_offset", since = "1.61.0")]
     #[inline(always)]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+
```
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:407]: Why are you introducing fresh blank lines (here and below)?

File: library/core/src/ptr/mut_ptr.rs [hunk 7]
```diff
@@ -1033,7 +1036,7 @@ impl<T: ?Sized> *mut T {
     )]
     // Postcondition: If `T` is a unit type (`size_of::<T>() == 0`), no allocation check is needed.
     // Otherwise, for non-unit types, ensure that `self` and `result` point to the same allocated object,
-    // verifying that the result remains within the same allocation as `self`.
+    // verifying that the result remains within the same allocation as `self`.
```
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:1039]: No need for adding whitespace.

File: library/core/src/ptr/mut_ptr.rs (hunk 3 above)
Comment (tautschnig) [reply library/core/src/ptr/mut_ptr.rs:2537]: Still adds whitespace at the end of the line.

File: library/core/src/ptr/mut_ptr.rs [hunk 8]
```diff
@@ -2422,7 +2502,7 @@ mod verify {
     generate_mut_arithmetic_harness!(u32, check_mut_add_u32, add);
     generate_mut_arithmetic_harness!(u64, check_mut_add_u64, add);
     generate_mut_arithmetic_harness!(u128, check_mut_add_u128, add);
-    generate_mut_arithmetic_harness!(usize, check_mut_add_usize, add);
+    generate_mut_arithmetic_harness!(usize, check_mut_add_usize, add);
```
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:2505]: Adding whitespace.

File: library/core/src/ptr/mut_ptr.rs (hunk 7 above)
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs (hunk 8 above)
Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Removed in newest commit.

File: library/core/src/ptr/mut_ptr.rs [hunk 9]
```diff
@@ -2361,6 +2361,78 @@ impl<T: ?Sized> PartialOrd for *mut T {
 #[unstable(feature = "kani", issue = "none")]
 mod verify {
     use crate::kani;
+    use core::mem;
+    // Constant for array size used in all tests
+    const ARRAY_SIZE: usize = 5;
```
Comment (tautschnig) [inline library/core/src/ptr/mut_ptr.rs:2366]: Could you please add a further comment saying _why_ you have chosen this particular value?

Comment (szlee118) [reply library/core/src/ptr/mut_ptr.rs]: Comment added.

**Comment by feliperodri** [comment]
@tautschnig @carolynzech could you do another round of review here? We need one more approval before merging it.

**Pull request merged**